src/main/java/org/springframework/data/repository/config/RepositoryBeanNameGenerator.java

```diff
@@ -15,56 +15,68 @@
 import org.springframework.beans.factory.config.BeanDefinition;
 import org.springframework.beans.factory.support.BeanDefinitionRegistry;
 import org.springframework.beans.factory.support.BeanNameGenerator;
 import org.springframework.context.annotation.AnnotationBeanNameGenerator;
+import org.springframework.context.annotation.ScannedGenericBeanDefinition;
 import org.springframework.util.ClassUtils;
 
 /**
  * Special {@link BeanNameGenerator} to create bean names for Spring Data repositories. Will delegate to an
  * {@link AnnotationBeanNameGenerator} but let the delegate work with a customized {@link BeanDefinition} to make sure
  * the repository interface is inspected and not the actual bean definition class.
- * 
+ *
  * @author Oliver Gierke
+ * @author Jens Schauder
  */
-public class RepositoryBeanNameGenerator implements BeanNameGenerator, BeanClassLoaderAware {
+public class RepositoryBeanNameGenerator {
 
 	private static final BeanNameGenerator DELEGATE = new AnnotationBeanNameGenerator();
 
-	private ClassLoader beanClassLoader;
+	private final ClassLoader beanClassLoader;
 
-	/* 
-	 * (non-Javadoc)
-	 * @see org.springframework.beans.factory.BeanClassLoaderAware#setBeanClassLoader(java.lang.ClassLoader)
-	 */
-	public void setBeanClassLoader(ClassLoader classLoader) {
-		this.beanClassLoader = classLoader;
+
+	public RepositoryBeanNameGenerator(ClassLoader beanClassLoader) {
+		this.beanClassLoader = beanClassLoader;
 	}
 
-	/* 
-	 * (non-Javadoc)
-	 * @see org.springframework.beans.factory.support.BeanNameGenerator#generateBeanName(org.springframework.beans.factory.config.BeanDefinition, org.springframework.beans.factory.support.BeanDefinitionRegistry)
+	/**
+	 * Generate a bean name for the given bean definition.
+	 *
+	 * @param definition the bean definition to generate a name for
+	 * @return the generated bean name
+	 * @since 2.0
 	 */
-	public String generateBeanName(BeanDefinition definition, BeanDefinitionRegistry registry) {
+	public String generateBeanName(BeanDefinition definition) {
 
 		AnnotatedBeanDefinition beanDefinition = new AnnotatedGenericBeanDefinition(getRepositoryInterfaceFrom(definition));
-		return DELEGATE.generateBeanName(beanDefinition, registry);
+		return DELEGATE.generateBeanName(beanDefinition, null);
 	}
 
 	/**
 	 * Returns the type configured for the {@code repositoryInterface} property of the given bean definition. Uses a
 	 * potential {@link Class} being configured as is or tries to load a class with the given value's {@link #toString()}
 	 * representation.
-	 * 
+	 *
 	 * @param beanDefinition
 	 * @return
 	 */
 	private Class<?> getRepositoryInterfaceFrom(BeanDefinition beanDefinition) {
+		if (beanDefinition instanceof ScannedGenericBeanDefinition) {
+			try {
+				return ((ScannedGenericBeanDefinition) beanDefinition).resolveBeanClass(beanClassLoader);
+			} catch (ClassNotFoundException e) {
+				throw new IllegalStateException("Could not resolve bean class.", e);
+			}
+		} else {
```

Comment:
We don't need an else-block here as the first branch always yields a result, right?